README
FreeRunner Rotate
Author:  Rui Miguel Silva Seabra <rms@1407.org>
License: GPLv3+.
Description: Inspired upon Chris Ball <cjb@laptop.org>'s previous
	     incarnation of rotate, this is a rewritten-from-scratch
	     new software that does some simple actions according to
	     the position detected by the FreeRunner's accelerometers.


######################
# How to make it run #
######################

1. Just start your phone
2. If it dies for some reason (like, currently, suspending), click on the nifty icon


ChangeLog
2010-05-01 - 0.5.8
	* Integrate Tim Abell's improvements to rotation sensitivity
	* Be smarter about the brightness paths, which change in 2.6.32
	* A few fixes here and there.

2010-04-04 - 0.5.7
	* Proper release.

2010-04-04 - 0.5.5
	* Allow disabling autostart through omnewrotate.conf
	* Allow different paths for brightness and accelerometers
	* Integrate Tim Abell's improvements to startup scripts
	* Get rid of gnome-autogen (not really needed here)

2009-05-02 - 0.5.4
	* Release 0.5.4 with fix for issues 5 and 6

2008-12-26 - 0.5.3
	* fix issue 4, for the time being frameworkd-glib isn't really used
	* added a compile time option for building without frameworkd-glib to
	  support distributions which don't use it.

2008-12-09 - 0.5.2
	* fix issues 1 and 2

2008-12-09 - 0.5.1 - Automagically lazy
	* same as 0.5.0 plus auto tools magic
	* desktop icon
	* places a starter script

2008-11-19 - 0.5.0 - Lazy edition
	* uses a second thread for reading the accellerometer packets
	* drops Fabian's changes (not that they weren't good, just not
	  needed any more)
	* adds flags (look at display_help() or ./rotate -h)
	* drops packets with 0 value coordinates (I got bogus packets
	  like that so I decided to drop them, if you feel you get good
	  packets with 0 value, you can use '-0' as an argument to take
	  them in account).
	* top -d 1 -p <PID> shows 0.0% CPU usage (of omnewrotate) even
	  during rotation
	* seems to waste a little too much memory (some stuff could be
	  done with one number and bitwise operations instead of several
	  numbers, I wonder how much that will improve and if it's worth
	  the effort...)
	* only output outside of debug mode are real errors

2008-10-14 - 0.3.0
	* dims while rotating and doesn't rotate if dimmed, sleeping for 5s

2008-09-22 - 0.2.1
	* more heuristic fixes (hopefully really fix when laying around, and
	  turned up)
	* workaround with alarm() for the accelerometer read-hang problem
	* sleeping for 100ms seems to get cpu usage quite low (0% to 0.9%)

2008-09-22 - 0.2.0
	* greatly improves heuristics
	* fixed heuristics when laying around, and turned up

2008-09-21 - 0.1.1 - First brown paper bag release
	* improves heuristics

2008-09-21 - 0.1.0 - First release.
	* makes some rotations


test.c
/*
 * Copyright © 2008 Rui Miguel Silva Seabra <rms@1407.org>
 * Copyright © 2008 Fabian Henze <hoacha@quantentunnel.de>
 *
 * Inspired upon Chris Ball's rotate, this is a totally new rewrite.
 *
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <signal.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <fcntl.h>
#include <linux/input.h>
#include <X11/Xlib.h>
#include <X11/extensions/Xrandr.h>

void packet_loop();
int read_packet(int from, int *x, int *y, int *z);
int get_position(int x, int y, int z);
void rotate(int pos);
void catch_alarm(int var);
void set_options(char **args);

/* we're reading this kind of events:
struct input_event {
    struct timeval time;
    __u16 code;
    __u16 type;
    __s32 value;
};
*/

#define UPRIGHT 0
#define RIGHT 1
#define UPSIDEDOWN 2
#define LEFT 3
#define EVENT_PATH "/dev/input/event3"

int current_x, current_y, current_z;
int current_pos = -1;
int file = -1;

static Display *display;
static Window rootWindow;
ushort debug = 0;
ushort change_brightness = 0;


#define BIG_DIFFERENCE 400
/* this ----v idea doesn't seem to be working right, for later investigation */
#define LONG_TIME 0

#define EVENT_PATH "/dev/input/event3"
#define GET_BRIGHTNESS_PATH "/sys/class/backlight/pcf50633-bl/actual_brightness"
#define SET_BRIGHTNESS_PATH "/sys/class/backlight/pcf50633-bl/brightness"



int set_brightness_file = -1;
int get_brightness_file = -1;

void *packet = NULL;


int main(int argc, char **argv)
{
    int x, y, z;

    file = open(EVENT_PATH, O_RDONLY);
    if (file < 0) {
	fprintf(stderr, "Can't open '%s': %s\n", EVENT_PATH,
		strerror(errno));
	exit(1);
    }

    set_options(argv);

    if (change_brightness) {
	set_brightness_file = open(SET_BRIGHTNESS_PATH, O_RDWR);
	if (set_brightness_file < 0) {
	    fprintf(stderr, "Can't open '%s': %s\n", SET_BRIGHTNESS_PATH,
		    strerror(errno));
	    fprintf(stderr, "No brightness control enabled\n");
	}
    }


    /* initialize current position */
    while (!read_packet(file, &x, &y, &z));
    current_pos = get_position(x, y, z);

    display = XOpenDisplay(":0");
    if (display == NULL) {
	fprintf(stderr, "Can't open display %s\n", XDisplayName(":0"));
	exit(1);
    }

    packet_loop();
}

void set_options(char **args)
{
    char *arg = (char *) *args;
    int i = 1;

    printf("OPTIONS:\n");
    while (arg != NULL) {
	if (0 == strncmp(arg, "debug", 5)) {
	    printf("RUNNING IN DEBUG MODE\n");
	    debug = 1;
	} else if (0 == strncmp(arg, "brightness", 10)) {
	    if (debug)
		printf("USING BRIGHT CONTROL\n");
	    change_brightness = 1;
	}
	arg = (char *) *(args + i++);
    }
}

void packet_loop()
{
    int x, y, z;
    int pos, last_pos = -1;
    char *orientation, *current_orientation;

    while (1) {
	    while (!read_packet(file, &x, &y, &z));
		printf("v(x,y,z)=(% 5d,% 5d,% 5d)\n",x,y,z);
    }
}

int read_packet(int from, int *x, int *y, int *z) {
    static struct input_event event_x, event_y, event_z, event_syn;
    void *packet_memcpy_result = NULL;
    int packet_size = sizeof(struct input_event);
    int size_of_packet = 4 * packet_size;
    int bytes_read = 0;

    signal(SIGALRM, catch_alarm);
    alarm(5);

    packet = malloc(size_of_packet);

    if (!packet) {
		fprintf(stderr, "malloc failed\n");
		exit(1);
    }

    bytes_read = read(from, packet, size_of_packet);

    if (bytes_read < packet_size) {
		fprintf(stderr, "fread failed\n");
		exit(1);
    }

    /* obtain the full packet */
    packet_memcpy_result = memcpy(&event_x,   packet,                   packet_size);
    packet_memcpy_result = memcpy(&event_y,   packet +     packet_size, packet_size);
    packet_memcpy_result = memcpy(&event_z,   packet + 2 * packet_size, packet_size);
    packet_memcpy_result = memcpy(&event_syn, packet + 3 * packet_size, packet_size);

    *x = event_x.value;
    *y = event_y.value;
    *z = event_z.value;

    free(packet);

    packet = NULL;
    signal(SIGALRM, SIG_DFL);
    alarm(0);

    if (event_syn.type == EV_SYN && *x != 0 && *y != 0 && *z != 0) {
		printf("Good packet: ");
		return (1);
	} else {
		printf("Bad packet!\n");
		return (0);
	}
}

ushort neighbour(int value, int target, int neighbour)
{
    return (target - abs(neighbour) < value
	    && value <= target + abs(neighbour));
}

int get_position(int x, int y, int z)
{

    const int total_accel = x ^ 2 + y ^ 2 + z ^ 2;

    if ((900 <= abs(z)) && (abs(z) < 1100)) {
	if (debug)
	    printf(" low angle\n");
	return (-1);
    }

    /* if ((900000 > total_accel) || (total_accel > 1200000)) {
       if(debug) printf(" shaking around\n");
       return(-2);
       } */

    /* if (900000 > total_accel) {
       if(debug) printf(" too little shaking\n");
       return(-2);
       } */

    if (total_accel > 1200000) {
	if (debug)
	    printf(" too much shaking\n");
	return (-2);
    }

    switch (current_pos) {
    case UPRIGHT:
	if (y < 0 && abs(x) < -2 * y)
	    return (UPRIGHT);
	else if (y > 0 && abs(x) < y)
	    return (UPSIDEDOWN);
	else if (x > 0)
	    return (RIGHT);
	else if (x < 0)
	    return (LEFT);
	break;
    case RIGHT:
	if (x > 0 && abs(y) < 2 * x)
	    return (RIGHT);
	else if (x < 0 && abs(y) < -x)
	    return (LEFT);
	else if (y > 0)
	    return (UPSIDEDOWN);
	else if (y < 0)
	    return (UPRIGHT);
	break;
    case LEFT:
	if (x < 0 && abs(y) < -2 * x)
	    return (LEFT);
	else if (x > 0 && abs(y) < x)
	    return (RIGHT);
	else if (y > 0)
	    return (UPSIDEDOWN);
	else if (y < 0)
	    return (UPRIGHT);
	break;
    case UPSIDEDOWN:
	if (y > 0 && abs(x) < 2 * y)
	    return (UPSIDEDOWN);
	else if (y < 0 && abs(x) < -y)
	    return (UPRIGHT);
	else if (x > 0)
	    return (RIGHT);
	else if (x < 0)
	    return (LEFT);
	break;
    default:
	if (y < 0 && abs(x) < -y)
	    return (UPRIGHT);
	else if (y > 0 && abs(x) < y)
	    return (UPSIDEDOWN);
	else if (x > 0)
	    return (RIGHT);
	else if (x < 0)
	    return (LEFT);
	break;
    }

    return (-1);
}

void rotate(int pos)
{
    ushort return_val = -1;

    Rotation r_to, r;
    XRRScreenConfiguration *config;
    int current_size = 0;
    int screen = -1;
    char current_brightness[3] = "63\n";
    char brightness_off[2] = "0\n";
    ushort do_rotate = 0;

    screen = DefaultScreen(display);
    rootWindow = RootWindow(display, screen);
    XRRRotations(display, screen, &r);

    switch (pos) {
    case UPRIGHT:
	r_to = RR_Rotate_0;
	do_rotate = 1;
	break;
    case RIGHT:
	r_to = RR_Rotate_90;
	do_rotate = 1;
	break;
    case UPSIDEDOWN:
	r_to = RR_Rotate_180;
	do_rotate = 1;
	break;
    case LEFT:
	r_to = RR_Rotate_270;
	do_rotate = 1;
	break;
    default:
	break;
    }

    if (do_rotate) {
	if (change_brightness && set_brightness_file >= 0) {
	    lseek(get_brightness_file, 0, SEEK_SET);
	    read(get_brightness_file, &current_brightness, 2);
	    lseek(set_brightness_file, 0, SEEK_SET);
	    write(set_brightness_file, &brightness_off, 2);
	    usleep(500000);
	}

	if (debug)
	    printf("ROTATING!\n");
	config = XRRGetScreenInfo(display, rootWindow);
	current_size = XRRConfigCurrentConfiguration(config, &r);
	XRRSetScreenConfig(display, config, rootWindow, current_size, r_to,
			   CurrentTime);
	usleep(500000);

	if (change_brightness && set_brightness_file >= 0) {
	    lseek(set_brightness_file, 0, SEEK_SET);
	    write(set_brightness_file, &current_brightness, 3);
	}
	current_pos = pos;
    }
}

int screen_dimmed()
{
    char current_brightness[3] = "63\n";
    short currently_bright = 0;

    if (!change_brightness)
	return (0);

    get_brightness_file = open(GET_BRIGHTNESS_PATH, O_RDONLY);
    if (get_brightness_file < 0) {
	fprintf(stderr, "Can't open '%s': %s\n", GET_BRIGHTNESS_PATH,
		strerror(errno));
	fprintf(stderr, "No brightness check enabled\n");
	return (0);
    } else {
	/* read the current brightness, and dim only if it's bright */
	lseek(get_brightness_file, 0, SEEK_SET);
	read(get_brightness_file, &current_brightness, 2);
	currently_bright = atoi(current_brightness);

	close(get_brightness_file);

	if (debug)
	    printf("Dimmed level: %d\n", currently_bright);

	if (currently_bright > 0)
	    return (0);
	else
	    return (1);

    }
}

int screen_locked()
{
    /* shouldn't rotate if screen is locked, waste of energy */
    return (0);
}

void catch_alarm(int var)
{
    if (packet) {
	free(packet);
	packet = NULL;
    }
    packet_loop();
}


gen-src-archive.sh
#!/bin/bash
release=$1
if [ "$release" == "" ]; then
	release="master"
	release=`git log -1 --pretty=format:%h $release`
	git archive --prefix omnewrotate-$release/ $release | gzip > omnewrotate-$release.tar.gz
else
	git archive --prefix omnewrotate-$release/ $release > omnewrotate-$release.tar
	mkdir fixup-tmp
	tar -C fixup-tmp -xf omnewrotate-$release.tar omnewrotate-$release/setup.py
	sed -i s/##version##/\"$release\"/ fixup-tmp/omnewrotate-$release/setup.py
	tar --delete -f omnewrotate-$release.tar omnewrotate-$release/setup.py
	tar -C fixup-tmp -uf omnewrotate-$release.tar omnewrotate-$release/setup.py
	rm -rf fixup-tmp
	gzip omnewrotate-$release.tar
	md5sum omnewrotate-$release.tar.gz > omnewrotate-$release.tar.gz.md5
fi
echo "archive omnewrotate-$release.tgz created. contents:"
tar -tzf omnewrotate-$release.tar.gz


src/omnewrotate.c
/* vim:tabstop=4
 * Copyright © 2008 Rui Miguel Silva Seabra <rms@1407.org>
 *
 * Inspired upon Chris Ball's rotate, this is a totally new rewrite.
 *
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <config.h>

#include <signal.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <fcntl.h>
#include <linux/input.h>
#include <X11/Xlib.h>
#include <X11/extensions/Xrandr.h>
#include <pthread.h>

#ifdef HAVE_LIBFRAMEWORD_GLIB

#include <dbus/dbus.h>

#endif

void *packet_reading_thread(void *);
void define_position(void);
int read_packet(void);
int neighbour(int value, int target, int neighbour);
void do_rotation(void);
void display_version(void);
void display_help(void);


/* we're reading this kind of events:
struct input_event {
	struct timeval time;
	__u16 code;
	__u16 type;
	__s32 value;
};
*/

#define UP 0
#define RIGHT 1
#define DOWN 2
#define LEFT 3

int x = 0;
int y = 0;
int z = 0;

/* Position flags */
int face_up = 1;
int vertical = 0;
int left = 0;
int right = 0;
int up = 0;
int down = 0;

int current_pos = -1;
int event3 = -1;
int read_sleep = 250000;
int accel_threshold = 50; //how close to 1000 in x/y/z accelerometer needs to read before rotating (command line option "-t")
int kernel_version = 2629; // 2629 == 2.6.29 (default), 2632 = 2.6.32

static Display *display;
static Window rootWindow;
ushort debug = 0;
ushort skip_zero = 1;
ushort change_brightness = 0;
#ifdef HAVE_LIBFRAMEWORD_GLIB
ushort use_dbus = 1;
#else
ushort use_dbus = 0;
#endif


#define BIG_DIFFERENCE 400
/* this ----v idea doesn't seem to be working right, for later investigation */
#define LONG_TIME 0

#define EVENT_PATH "/dev/input/event3"
#define GET_BRIGHTNESS_PATH_2629 "/sys/class/backlight/gta02-bl/actual_brightness"
#define SET_BRIGHTNESS_PATH_2629 "/sys/class/backlight/gta02-bl/brightness"
#define GET_BRIGHTNESS_PATH_2632 "/sys/class/backlight/pcf50633-bl/actual_brightness"
#define SET_BRIGHTNESS_PATH_2632 "/sys/class/backlight/pcf50633-bl/brightness"

#define NUM_THREADS 1


int set_brightness_file = -1;
int get_brightness_file = -1;
char current_brightness[4];

void set_linux_type(int kernel_version, char *set_brightness_path, char *get_brightness_path)
{
	struct stat buf;

	switch(kernel_version)
	{
		case 0:	// unspecified guessing
		{
			if(! (set_brightness_path && get_brightness_path) )
			{
				fprintf(stderr, "With unspecified Linux version, you need to define set_brightness and get_brightness paths.\n");
				exit(1);
			}
		}
		case 2632:	// Linux 2.6.32
		{
			if(set_brightness_path)
				free(set_brightness_path);
			if(!get_brightness_path)
				free(get_brightness_path);
			if((stat(SET_BRIGHTNESS_PATH_2632, &buf) == 0 && stat(GET_BRIGHTNESS_PATH_2632, &buf) == 0))
			{
				set_brightness_path = SET_BRIGHTNESS_PATH_2632;
				get_brightness_path = GET_BRIGHTNESS_PATH_2632;
				snprintf(current_brightness, 5, "63\n");
			}
			else
			{
				if((stat(SET_BRIGHTNESS_PATH_2629, &buf) == 0 && stat(GET_BRIGHTNESS_PATH_2629, &buf) == 0))
				{
					fprintf(stderr, "Expected Linux 2.6.32 style but found Linux 2.6.29 style paths for brightness\n");
					set_brightness_path = SET_BRIGHTNESS_PATH_2629;
					get_brightness_path = GET_BRIGHTNESS_PATH_2629;
					snprintf(current_brightness, 5, "255\n");
				}
				else
				{
					fprintf(stderr, "Unsupported Linux version\n");
					exit(2);
				}
			}
			break;
		}
		default:	// Linux 2.6.29
		{
			if(set_brightness_path)
				free(set_brightness_path);
			if(get_brightness_path)
				free(get_brightness_path);
			if((stat(SET_BRIGHTNESS_PATH_2629, &buf) == 0 && stat(GET_BRIGHTNESS_PATH_2629, &buf) == 0))
			{
				set_brightness_path = SET_BRIGHTNESS_PATH_2629;
				get_brightness_path = GET_BRIGHTNESS_PATH_2629;
				snprintf(current_brightness, 5, "255\n");
			}
			else
			{
				if((stat(SET_BRIGHTNESS_PATH_2632, &buf) == 0 && stat(GET_BRIGHTNESS_PATH_2632, &buf) == 0))
				{
					fprintf(stderr, "Expected Linux 2.6.29 style but found Linux 2.6.32 style paths for brightness\n");
					set_brightness_path = SET_BRIGHTNESS_PATH_2632;
					get_brightness_path = GET_BRIGHTNESS_PATH_2632;
					snprintf(current_brightness, 4, "63\n");
				}
				else
				{
					fprintf(stderr, "Unsupported Linux version\n");
					exit(2);
				}
			}
			break;
		}
	}
}

int main(int argc, char **argv)
{
	struct stat buf;
	pthread_t p_thread[1];
	int i;
	int pos=0;
	static char options[] = "pbd0hva:g:s:t:k:";
	int option;
	char * accelerometer = (char*)NULL;
	char * get_brightness_path = (char*)NULL;
	char * set_brightness_path = (char*)NULL;

#ifdef HAVE_LIBFRAMEWORD_GLIB
	DBusError err;
	DBusConnection* conn;
#endif
	int ret=0;

	while((option = getopt(argc,argv,options)) != -1)
	{
		switch(option)
		{
			case '0':
			{
				skip_zero=0;
				break;
			}
			case 'a':
			{
				accelerometer = strndup(optarg, 1024);
				if(stat(accelerometer, &buf) < 0)
				{
					fprintf(stderr, "Problem accessing accelerometer at %s: %s\n", accelerometer, strerror(errno));
					free(accelerometer);
					exit(1);
				}
				break;
			}
			case 'b':
			{
				change_brightness=1;
				break;
			}
			case 'd':
			{
				debug=1;
				break;
			}
			case 'g':
			{
				get_brightness_path = strndup(optarg, 1024);
				if(stat(get_brightness_path, &buf) < 0)
				{
					free(get_brightness_path);
					get_brightness_path=NULL;
				}
				break;
			}
			case 'h':
			{
				display_help();
				exit(0);
			}
			case 'p':
			{
				read_sleep = 1000000;
				break;
			}
			case 's':
			{
				set_brightness_path = strndup(optarg, 1024);
				if(stat(set_brightness_path, &buf) < 0)
				{
					free(set_brightness_path);
					set_brightness_path=NULL;
				}
				break;
			}
			case 'k':
			{
				kernel_version = atoi(optarg);
				break;
			}
			case 't':
			{
				accel_threshold = atoi(optarg);
				break;
			}
			case 'v':
			{
				display_version();
				exit(0);
			}
			default:
			{
				exit(1);
			}
		}
	}

	set_linux_type(kernel_version, set_brightness_path, set_brightness_path);

	display = XOpenDisplay(":0");
	if (display == NULL)
	{
		fprintf(stderr, "Can't open display %s\n", XDisplayName(":0"));
		exit(1);
	}

#ifdef HAVE_LIBFRAMEWORD_GLIB
	dbus_error_init(&err);
	conn = dbus_bus_get(DBUS_BUS_SESSION, &err);
	if (dbus_error_is_set(&err))
	{
		fprintf(stderr, "Connection Error (%s), so not using dbus\n", err.message);
		dbus_error_free(&err);
	}
	if (NULL == conn) use_dbus=0;
#endif

	if (change_brightness && !use_dbus) {
		set_brightness_file = open(set_brightness_path, O_WRONLY);
		get_brightness_file = open(get_brightness_path, O_WRONLY);

		if (set_brightness_file < 0 || get_brightness_file < 0)
		{
			fprintf(stderr, "Can't open '%s': %s\n", set_brightness_path, strerror(errno));
			fprintf(stderr, "No brightness control enabled\n");
			change_brightness = 0;
		}
	}

	if(accelerometer == NULL) event3 = open(EVENT_PATH, O_RDONLY);
	else event3 = open(accelerometer, O_RDONLY);

	if (event3 < 0)
	{
		fprintf(stderr, "Can't open '%s': %s\n", EVENT_PATH,
		strerror(errno));
		exit(1);
	}

	/* Create a the packet reading thread */
	for(i = 0; i < NUM_THREADS; i++)
	{
		if(pthread_create(&p_thread[i], NULL, packet_reading_thread, (void *)(long)i) != 0)
		{
			fprintf(stderr, "Error creating the packet reading thread");
			exit(1);
		}
	}

	if (debug) printf("Begin loop.\n");
	while(1)
	{
		usleep(read_sleep);
		pos=current_pos;
		define_position();
		if(current_pos != pos) do_rotation();
	}
	return(ret);
}

void display_version(void)
{
	printf("omnewrotate version %s is licensed under the GNU GPLv3 or later.\n", VERSION);
}

void display_help(void)
{
	display_version();
	printf("\nUsage:\n"\
		"	-0	Don't skip packets if any coordinate is '0'\n"
		"	-a	Accelerometer path, by default '" EVENT_PATH "'\n"
		"	-b	Use brightness (dimming and back) effects\n"
		"	-d	Debug mode (extra yummy output)\n"
		"	-k	Set the kernel version. Use 2629 (default, Linux 2.6.29)\n"
		"       or 2632 to change to Linux 2.6.32 or greater\n"
		"       use 0 to be able to arbitrarily change brightness paths\n"
		"	-g	Get current brightness path,\n"
		"		by default '" GET_BRIGHTNESS_PATH_2629 "'\n"
		"	-h	Help (what you're reading right now)\n"
		"	-p	Powersaving features (like sleeping longer, etc...)\n"
		"	-s	Set current brightness path,\n"
		"		by default '" SET_BRIGHTNESS_PATH_2629 "'\n"
		"	-t	Threshold for rotating. Default +/-50 (approx 20deg) \n"
		"		against target of 1000. 300 gives approx 45deg\n"
		"	-v	Show version and license\n"
		"\n"
		);
}

void do_rotation(void)
{
	Rotation r_to = RR_Rotate_0;
	Rotation r = RR_Rotate_0;
	XRRScreenConfiguration *config;
	int screen = -1;
	int current_size = 0;

	screen = DefaultScreen(display);
	rootWindow = RootWindow(display, screen);
	XRRRotations(display, screen, &r);

	char brightness_off[2] = "0\n";

	switch(current_pos)
	{
		case(UP):
			{
				if(debug) printf("Rotating to UP\n");
				r_to = RR_Rotate_0;
				break;
			}
		case(DOWN):
			{
				if(debug) printf("Rotating to DOWN\n");
				r_to = RR_Rotate_180;
				break;
			}
		case(RIGHT):
			{
				if(debug) printf("Rotating to RIGHT\n");
				r_to = RR_Rotate_90;
				break;
			}
		case(LEFT):
			{
				if(debug) printf("Rotating to LEFT\n");
				r_to = RR_Rotate_270;
				break;
			}
		default: break;
	}

	if (change_brightness)
	{
		if(debug) printf("Dimming screen for nifty effect\n");

		if(use_dbus)
		{
		}
		else
		{
			lseek(get_brightness_file, 0, SEEK_SET);
			read(get_brightness_file, &current_brightness, 3);
			lseek(set_brightness_file, 0, SEEK_SET);
			write(set_brightness_file, &brightness_off, 2);
		}
		usleep(500000);
	}

	config = XRRGetScreenInfo(display, rootWindow);
	current_size = XRRConfigCurrentConfiguration(config, &r);
	XRRSetScreenConfig(display, config, rootWindow, current_size, r_to, CurrentTime);

	if (change_brightness)
	{
		if(debug) printf("Recovering screen brightness for nifty effect\n");
		usleep(500000);
		lseek(set_brightness_file, 0, SEEK_SET);
		write(set_brightness_file, &current_brightness, 4);
	}

}

void define_position(void)
{

	/* Conclude all facts about current position */
	if(neighbour(z,0,500))
		vertical=1;
	else
		vertical=0;

	if(z>=0)
		face_up=1;
	else
		face_up=0;

	if(neighbour(x,-1000,accel_threshold))
		left=1;
	else
		left=0;

	if(neighbour(x,1000,accel_threshold))
		right=1;
	else
		right=0;

	if(neighbour(y,1000,accel_threshold))
		down=1;
	else
		down=0;

	if(neighbour(y,-1000,accel_threshold))
		up=1;
	else
		up=0;

	if(down) current_pos=DOWN;
	else if(up) current_pos=UP;
	else if(left) current_pos=LEFT;
	else if(right) current_pos=RIGHT;

	if(debug)
	{
		printf("v(x,y,z)=(% 5d,% 5d,% 5d) == ",x,y,z);

		printf(" face %s", face_up?"up":"down");
		printf(" %s-ish", vertical?"vertical":"horizontal");
		if(left) printf(" left");
		if(right) printf(" right");
		if(up) printf(" up");
		if(down) printf(" down");

		printf("\n");
	}
}

/**
  * neighbour(): test proximity of one value to another based on a threshold.
  * returns true if "value" is nearer to "target" than threshold "value"
  * value: the varying value to test
  * target: the number to test proximity of to
  * threshold: the threshold for how close "value" has to be to "target" to succeed
  * eg: for target=100 and threshold=10, values of 91 to 110 inclusive would return true
  */
int neighbour(int value, int target, int threshold)
{
	return (value > (target - abs(threshold)) && (value <= (target + abs(threshold))));
}

void *packet_reading_thread(void *arg)
{
	while(1)
	{
		while(!read_packet());
		usleep(250000);
	}
}

int read_packet()
{
	static struct input_event event_x, event_y, event_z, event_syn;
	void *packet_memcpy_result = NULL;
	int packet_size = sizeof(struct input_event);
	int size_of_packet = 4 * packet_size;
	int bytes_read = 0;
	void *packet = NULL;


	packet = malloc(size_of_packet);

	if (!packet)
	{
		fprintf(stderr, "malloc failed\n");
		exit(1);
	}

	bytes_read = read(event3, packet, size_of_packet);

	if (bytes_read < packet_size)
	{
		fprintf(stderr, "fread failed\n");
		exit(1);
	}

	/* obtain the full packet */
	packet_memcpy_result = memcpy(&event_x,   packet,                   packet_size);
	packet_memcpy_result = memcpy(&event_y,   packet +     packet_size, packet_size);
	packet_memcpy_result = memcpy(&event_z,   packet + 2 * packet_size, packet_size);
	packet_memcpy_result = memcpy(&event_syn, packet + 3 * packet_size, packet_size);

	free(packet);

	if(skip_zero && (event_x.value == 0 || event_y.value == 0 || event_z.value == 0))
	{
		if(debug) printf("Bad packet!\n");
		return(0);
	}

	if (event_syn.type == EV_SYN)
	{
		x = event_x.value;
		y = event_y.value;
		z = event_z.value;

		return (1);
	}
	else
		return (0);
}

